### ISSUE TITLE
VSCode not to associate itself with any file type unless the user chooses so during installation

### ISSUE THREAD

**Issue by darkred** [Microsoft/vscode#8687]
- VSCode Version: 1.2.1
- OS Version: win10 x64

Steps to Reproduce:

1. Install VSCode
2. Open `Control Panel > Control Panel Home > Default Programs > Assosiate a file type or protocol with a specific program`:  
you'll notice that a hundred file types (in fact 101 !) are now associated with VSCode ([screenshot 1](https://i.imgur.com/ntouoyZ.jpg)).   
There was no such option to opt in during the installation  ([screenshot 2](https://i.imgur.com/pZHEvn5.png))  
neither there's a way to remove the file associations inside the program preferences.  
The only way remove the file associations is to uninstall VSCode altogether.

For reference, all other editors that I have installed: Sublime Text, Atom, Notepad++,  gVim, UltraEdit 
don't associate themselves with any file type.

---

So, my suggestion: 
VSCode not to associate itself with any file extension unless the user chooses so during installation. 
Even better, to be able to configure file type associations with VSCode (enable/disable each one) , inside the program's preferences.

**Comment by Tyriar** [comment]
I think there's already a checkbox on the installer for this. It's checked by default though, it should probably be opt-in.

**Comment by darkred** [comment]
No, there's no such checkbox. Please check [screenshot 2](https://i.imgur.com/pZHEvn5.png) from my 1st post, and the attached video capture of the install procedure: [Video_2016-07-03_235157.zip](https://github.com/Microsoft/vscode/files/345298/Video_2016-07-03_235157.zip)

Also, the two checkboxes in the installer (esp. the 1st one) to `Add "Open with Code" action to Windows Explorer file/directory context menu`are unchecked by default, and even so (i.e. keeping them unchecked), the associations are made in spite of that.

**Comment by joaomoreno** [comment]
My belief is that Code doesn't associate itself automatically, but simply registers as a possible editor for such files.

**Comment by darkred** [comment]
No, that's not the case.
Installer doesn't just add Code as an extra entry in windows explorer's `rightclick|Open with` menu for these file extensions:
![2016-07-04_183804](https://cloud.githubusercontent.com/assets/723651/16578936/ebfd06e0-42a5-11e6-99c5-9b4bc54a1b1a.png)

I created various empty files with random extensions from my [screenshot 1](https://i.imgur.com/ntouoyZ.jpg)) 
and **all are associated with Code** :
![2016-07-04_184313](https://cloud.githubusercontent.com/assets/723651/16578947/f906281c-42a5-11e6-80a7-d7dabfe05650.png)
In other words, doubleclicking any of these files, opens them in Code.

**Comment by joaomoreno** [comment]
Yes, but did you have **any other** application that was able to open those files before you installed Code?

**Comment by darkred** [comment]
The `.js` files open with `Microsoft Windows Based Script Host` by default.
Also, `sh` files

**Comment by darkred** [comment]
_(sorry, closed the issue by mistake)_

The `.js` files would open with `Microsoft Windows Based Script Host` (a windows component, it's by default).
After installing Code they become assosiated with Code.

On the other hand, `sh` files would open with `Git for Windows`. 
After installing Code they are still associated with Git.

**Comment by joaomoreno** [comment]
Got it, thanks for the info! I will attempt to reproduce on a brand new Windows box, see what we can do.

**Comment by joaomoreno** [comment]
**Investigation**

VSCode simply registers itself as an application suitable to open a variety of file types, through the use of the [`HKEY_CLASSES_ROOT`](https://technet.microsoft.com/en-us/library/dd670c1d-2501-4f32-885b-0c6a1ae662f4) registry subtree. Every time an application adds itself as one able to handle a file type, **Windows** will do the following:

- If the file type itself doesn't have an application association set by the user, it will pick VS Code since it is the only alternative available.
- Otherwise, it will **not** overwrite the association.

The exception to the rule seems to be files like `*.js` which seem to respect some strange _magic_ which I don't understand: they seem to be inherently associated with the `Microsoft Windows Based Script Host` through some other mechanism. As no one in their sane mind would actually expect to double click a `.js` file to run it in 2016, I'll let this slip. There may be others, but so far this is the one.

In conclusion, Code is doing the right thing, no change there.

**The Rabbit Hole goes deeper**

I've decided to experiment a bit more to find out how this works.

In a brand new VM, I installed Notepad++, created the following 4 files and associated `*.config` files to open with Notepad++.

![image](https://cloud.githubusercontent.com/assets/22350/17027915/25e52ed8-4f66-11e6-97b0-a3627a4818ef.png)

Note that:

- `*.txt` files are associated with `Notepad`.
- `*.sh` files are not associated with any application, since there is no suitable application that can handle them.
- `*.js` files are associated with the pesky `Microsoft Windows Based Script Host`.
- `*.config` files are associated with `Notepad++`.

After, I installed the latest Insiders setup:

![image](https://cloud.githubusercontent.com/assets/22350/17029107/2faee9be-4f6a-11e6-9c16-387ea308d58d.png)

Everything is the same. Wait, what..? But Code did get registered to a few file extensions. Why?

![image](https://cloud.githubusercontent.com/assets/22350/17029141/53b90862-4f6a-11e6-8bb4-2795bac496b7.png)

After investigating why that is, it happens to be the `PersistentHandler` registry key. That seems to be the difference between `*.sh` and `*.aspx` files for example:

![image](https://cloud.githubusercontent.com/assets/22350/17029270/e7e2b862-4f6a-11e6-9227-2d7b464f58bc.png)

![image](https://cloud.githubusercontent.com/assets/22350/17029266/e2c4328e-4f6a-11e6-85f5-855f67c45e15.png)

All the file extensions that were correctly associated with Code have this `PersistentHandler` set to some GUID, which was already there before Code's installation. I failed to understand where `PersistentHandler` comes from, and what its value actually means.

I stopped at this point, since I can think of more productive ways to spend my time.

**Conclusion**

I've added a checkbox during setup which will let users skip file associations: `Register Code as an editor for supported file types`.

@dbaeumer This is a new string that needs to be localised. The key is `AssociateWithFiles`, in `build\win32\i18n\messages.en.isl`.

**Comment by DarkNami** [comment]
VSCode Version: 1.4.0
OS Version: Win7 x64

After installing Visual Code, I don't know which file is .gitattributes, and which file is .gitignore ...
![filename](https://cloud.githubusercontent.com/assets/5100919/17802988/133568a8-6626-11e6-964f-6e1111714bf6.png)

although I added 'AlwaysShowExt' in filetype, but not working...
![regedit](https://cloud.githubusercontent.com/assets/5100919/17803025/43bf37c4-6626-11e6-875a-2772e4531a66.png)

**Comment by dbaeumer** [comment]
Assigning back to @joaomoreno to react to comments.

**Comment by DarkNami** [comment]
I uninstall and re-install again, uncheck "Register Code as an editor supported file types", it is ok now.
![2016-8-19 16-44-29](https://cloud.githubusercontent.com/assets/5100919/17804037/5a603220-662c-11e6-9bb8-f003f88e8872.png)

If I update from previous version (1.2.x), installer will associate filetypes.

Thanks.

**Comment by joaomoreno** [comment]
Thank you!

**Comment by laggingreflex** [comment]
Please add more options to set VSCode as default code editor, specially when switching from Sublime Text. 

Like in `Control Panel\Programs\Default Programs\Set Default Programs`

**Comment by cretace** [comment]
That box needs to spell out what it is going to do and it should show _every_ association so we can untick ones we don't wnat. It is worded too vaguely.
 The install removed my previously set up editor associations - they no longer appear under "open with". Very annoying.

**Comment by mrspeaker** [comment]
I was testing out VSCode for the last couple of days (it's great!) and just noticed it had changed a bunch of default file associations... which is not cool - if you "earn my trust" I'll make you default by choice. Anyhoo, for this I uninstalled it, and now I have a bunch of broken file associations... now I'm annoyed AND I have to google around how to fix it because my other editors-of-choice aren't so stupid as to overwrite them back.

I know you probably don't care, but just thought I'd say I think you should re-think this policy!

**Comment by trueneu** [comment]
...and it associates itself automatically with each and every file type it knows how to open on macOS. I love how VS Code thinks it is the only editor on my computer, but that's not the case.

**Comment by billinghamj** [comment]
How do I undo VSCode's damage on my Mac?

**Comment by rymcol** [comment]
I still have this issue with macOS as well

**Comment by mifi** [comment]
Solution: Delete VSCode or move it away from /Applications

**Comment by billinghamj** [comment]
Yeah in the end I just deleted it and switched back to Sublime Text
